pkg/algorithm/hived_algorithm.go

```diff
@@ -666,13 +605,125 @@ func (h *HivedAlgorithm) updateVCDoomedBadCells(c CellChain, l CellLevel) {
+			// even if some pods are now not within suggested nodes
+			klog.Warningf("Some nodes used by affinity group %v are no longer "+
+				"within K8s suggested nodes: %v", g.name, nodesNotInSuggested)
+		}
+		if podIndex = getNewPodIndex(g.allocatedPods[s.GpuNumber]); podIndex == -1 {
+			panic(internal.NewBadRequestError(fmt.Sprintf(
+				"Requesting more pods than the configured number for %v GPUs (%v pods) in affinity group %v",
+				s.GpuNumber, g.totalPodNums[s.GpuNumber], s.AffinityGroup.Name)))
+		}
+	} else { // groupPreempting
+		klog.Infof("[%v]: Pod affinity group is preempting others: %v",
+			internal.Key(pod), s.AffinityGroup.Name)
+		nodesNotInSuggested := collectNodesNotSuggested(g.physicalGpuPlacement, suggestedNodes)
+		if phase == internal.PreemptingPhase && !nodesNotInSuggested.IsEmpty() {
+			// If we find a preempting group's placement is not fully within suggested nodes, we should cancel
+			// the preemption so as to reschedule it to other places. We should do this only in Preempting phase
+			// because the suggested nodes in Filtering phase does not consider preemption.
+			klog.Infof("[%v]: Canceling affinity group %v's preemption because its placement is "+
+				"no longer fully within Preempting-phase suggested nodes (non-suggested nodes: %v)",
+				internal.Key(pod), g.name, nodesNotInSuggested)
+			h.deletePreemptingAffinityGroup(g, pod)
+		} else {
+			groupPhysicalPlacement = g.physicalGpuPlacement
+			groupVirtualPlacement = g.virtualGpuPlacement
+			if preemptionVictims, _ = collectPreemptionVictims(groupPhysicalPlacement); len(preemptionVictims) == 0 {
+				klog.Infof("Preemption victims have been cleaned up for the preemptor affinity group %v", g.name)
+			}
+			g.preemptingPods[pod.UID] = pod
+		}
+	}
+	return groupPhysicalPlacement, groupVirtualPlacement, preemptionVictims, podIndex
+}
+
+// schedulePodFromNewGroup schedules a pod from a new affinity group, find placement for the group,
+// and checks if the group needs preemption.
+func (h *HivedAlgorithm) schedulePodFromNewGroup(
+	s *api.PodSchedulingSpec,
+	suggestedNodes common.Set,
+	phase internal.SchedulingPhase,
+	pod *core.Pod) (
+	groupPhysicalPlacement groupPhysicalPlacement,
+	groupVirtualPlacement groupVirtualPlacement,
+	preemptionVictims map[string]common.Set,
+	waitReason string) {
+
+	groupPhysicalPlacement, groupVirtualPlacement, nodesNotInSuggested := h.scheduleNewAffinityGroup(
+		pod, s, suggestedNodes)
+	if groupPhysicalPlacement == nil {
+		if nodesNotInSuggested.IsEmpty() {
+			waitReason = "insufficient capacity in physical cluster"
+			if CellPriority(s.Priority) >= minGuaranteedPriority {
+				waitReason = fmt.Sprintf("insufficient capacity in VC %v", s.VirtualCluster)
+			}
+		} else {
+			// for an unallocated group,
+			// we will keep it waiting if not all of its pods are scheduled to suggested nodes
+			waitReason = fmt.Sprintf("affinity group is decided to be scheduled to some nodes "+
```

Comment:
>affinity group is decided to be scheduled to some nodes [](start = 29, length = 55)

"affinity group can only be scheduled to nodes not within K8s suggested nodes" ?